File: cpp/medium/ProductsofArrayExceptSelf.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  vector<int> productExceptSelf(vector<int> &nums) {
    int size = nums.size(), mult = 1;
    vector<int> prev(size), cur(size), ans(size);

    for (int i = 0; i < size; i++) {
      mult *= nums[i];
      prev[i] = mult;
    }
    mult = 1;

    for (int i = size - 1; i > 0; i--) {
      mult *= nums[i];
      cur[i] = mult;
    }
    mult = 1;

    ans[0] = cur[1];
    ans[size - 1] = prev[size - 2];

    int prevPointer = 0, curPointer = 2, anspointer = 1;

    while (curPointer < size) {
      ans[anspointer] = prev[prevPointer] * cur[curPointer];

      prevPointer++;
      curPointer++;
      anspointer++;
    }

    return ans;
  }
};

int main() {
  Solution *s1 = new Solution();
  vector<int> arr = {1, 2, 4, 6}, ans;
  ans = s1->productExceptSelf(arr);

  for (int i = 0; i < ans.size(); i++)
    cout << ans[i] << " ";
}

File: ts/medium/ProductsofArrayExceptSelf.ts

class Solution {
	productExceptSelf(nums: number[]): number[] {
		let preFix = new Array(nums.length + 1).fill(1);
		let postFix = new Array(nums.length + 1).fill(1);
		for (let idx = 1; idx < preFix.length; idx++) {
			preFix[idx] = preFix[idx - 1] * nums[idx - 1];
		}
		for (let idx = postFix.length - 2; idx >= 0; idx--) {
			postFix[idx] = postFix[idx + 1] * nums[idx];
		}
		let result = [];
		for (let i = 0; i < nums.length; i++) {
			result.push(preFix[i] * postFix[i + 1]);
		}
		return result;

	}
}



